internal/controller/etcdopstask/reconcile_delete_test.go

```diff
@@ -0,0 +1,277 @@
+		expectedLastErrors    *druidv1alpha1.LastError
+		expectedPhase         *druidv1alpha1.OperationPhase
+	}{
+		{
+			name:          "Task not found",
+			task:          nil,
+			cleanupFailed: false,
+		},
+		{
+			name:           "Task is in rejected state, last operation is updated",
+			task:           newTestTask(ptr.To(druidv1alpha1.TaskStateRejected)),
+			cleanupFailed:  false,
+			expectedResult: ctrlutils.ContinueReconcile(),
+			expectedLastOperation: &druidv1alpha1.LastOperation{
+				Type:        druidv1alpha1.LastOperationTypeDelete,
+				State:       druidv1alpha1.LastOperationStateProcessing,
+				Description: "Cleanup skipped for rejected task",
+			},
+			expectedPhase: ptr.To(druidv1alpha1.OperationPhaseCleanup),
+		},
+		{
+			name:           "result.requeue is false, no error, last operation is updated",
+			task:           newTestTask(nil),
+			cleanupFailed:  false,
+			expectedResult: ctrlutils.ContinueReconcile(),
+			expectedLastOperation: &druidv1alpha1.LastOperation{
+				Type:  druidv1alpha1.LastOperationTypeDelete,
+				State: druidv1alpha1.LastOperationStateProcessing,
+			},
+			expectedPhase: ptr.To(druidv1alpha1.OperationPhaseCleanup),
+		},
+		{
+			name:           "result.requeue is false, error, last operation and last Error is updated",
+			task:           newTestTask(nil),
+			expectedResult: ctrlutils.ReconcileWithError(druiderr.WrapError(fmt.Errorf("test error"), "TestError", "TestOperation", "This is a test error")),
+			cleanupFailed:  true,
+			resultRequeue:  false,
+			expectedLastOperation: &druidv1alpha1.LastOperation{
+				Type:  druidv1alpha1.LastOperationTypeDelete,
+				State: druidv1alpha1.LastOperationStateError,
+			},
+			expectedLastErrors: &druidv1alpha1.LastError{
+				Code:        druidv1alpha1.ErrorCode("TestError"),
+				Description: "This is a test error",
+			},
+			expectedPhase: ptr.To(druidv1alpha1.OperationPhaseCleanup),
+		},
+		{
+			name:           "result.requeue is true, no error, requeues task",
+			task:           newTestTask(nil),
+			cleanupFailed:  false,
+			resultRequeue:  true,
+			expectedResult: ctrlutils.ReconcileAfter(60*time.Second, "Task cleanup in progress"),
+			expectedPhase:  ptr.To(druidv1alpha1.OperationPhaseCleanup),
+		},
+		{
+			name:           "result.requeue is true, error, last Error is updated",
+			task:           newTestTask(nil),
+			expectedResult: ctrlutils.ReconcileWithError(druiderr.WrapError(fmt.Errorf("test error"), "TestError", "TestOperation", "This is a test error")),
+			cleanupFailed:  true,
+			resultRequeue:  true,
+			expectedLastErrors: &druidv1alpha1.LastError{
+				Code:        druidv1alpha1.ErrorCode("TestError"),
+				Description: "This is a test error",
+			},
+			expectedPhase: ptr.To(druidv1alpha1.OperationPhaseCleanup),
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			var cl client.Client
+			if tc.task != nil {
+				cl = setupFakeClient(tc.task, true)
```

Comment:
why are you using `setupFakeClient` which is also defined in a unit test file for `utils_test`. This should ideally move to test utility packages instead.